### DISCUSSION TITLE
seeing green check marks ✅ for failed github action runs

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2485] [Tap maintenance and Homebrew development]
why would i see a green check mark in a github action run, when the "pipeline" / build failed? shouldn't this be annotated with a red  X ❌ or something?

https://github.com/FreeCAD/homebrew-freecad/runs/4243010977?check_suite_focus=true

**Comment by carlocab** [comment]
That's working as designed (we don't want to see build failures for unbottled formulae in Homebrew/core). You do see annotations in the "Files" tab about the failure though. We can also tweak this behaviour a bit, but I'm not sure about the correct way to do this yet.

**Reply by mahrud** [reply]
I think it would be great if the warning explicitly said which formulae are not bottled on that particular architecture (when I run `brew unbottled` locally it says everything is bottled). Currently the warning is not very informative:
```
2022-12-15T03:51:27.3478550Z ##[warning]macaulay2 has unbottled dependencies, so a bottle will not be built.
```
This warning is from a github action running on macOS Catalina. Locally I run Linux, and I get this output:
```
$ brew unbottled --tag=catalina macaulay2
==> Populating dependency tree...
==> :catalina bottle status
macaulay2: already bottled
```
Probably because the previous bottle already exists. If I delete the bottle segment from the formula the command above works, but this one still mysteriously doesn't:
```
$ brew unbottled --tag=arm64_monterey M2
==> Populating dependency tree...
==> :arm64_monterey bottle status
macaulay2: doesn't support this macOS
```
(Note that this formula _has_ been bottled successfully for M1, so it definitely supports it!)

**Reply by SMillerDev** [reply]
It says which OS/arch reported that though. So adding it there won't give you more info.

**Reply by mahrud** [reply]
I don't follow. I was asking for something like this on `brew test-bot`:
```
##[warning] Formula has unbottled dependencies, so a bottle will not be built:
macaulay2: unbottled deps: cmake xz swig python@3.9 node texinfo node
```
Specifically because it is difficult/impossible to figure out which bottle is missing from a different architecture.

**Reply by SMillerDev** [reply]
But you know which architecture testbot is running on right? At least in Homebrew core the naming of the jobs include it: "CI / tests (11-arm64) (pull_request)"

**Reply by mahrud** [reply]
Oh yes, sure, I know the architecture where _some_ dependencies are unbotted. But unless I buy a machine of that architecture and run it there myself, I won't know which dependencies!

**Reply by SMillerDev** [reply]
I don't follow. If testbot reports it, then the architecture and OS of that run have unbottled dependencies.

**Reply by mahrud** [reply]
Yes, I know the architecture and OS, what I don't know is which formulae are unbottled on that architecture and OS!

**Reply by SMillerDev** [reply]
But that's what `brew unbottled` is for as you noticed. You can make a pull request to testbot to add this information there too, but I'm not sure what that would add since you'll need to do manual work to fix that test run anyway.

**Reply by mahrud** [reply]
Yes, I noticed this is what `brew unbottled` does, but there are two issues:
1. if a bottle for a given architecture has been created in the past and still exists in the formula file, `brew unbottled` doesn't actually check whether all dependencies are still currently bottled. e.g. `nest` has an existing bottle for Catalina, so you get that it's already bottled:
```
$ brew unbottled --tag=catalina nest
==> Populating dependency tree...
==> :catalina bottle status
nest: already bottled
```
Despite the fact that it depends on a formula that is not bottled for Catalina anymore:
```
$ brew deps nest | grep python
python@3.10
$ brew unbottled --tag=catalina python@3.10
==> Populating dependency tree...
==> :catalina bottle status
python@3.10: unbottled deps: xz
```
**The problem:** if I try to update `nest` or bump the revision, test-bot would only tell me that `nest` has unbottled dependencies on Catalina and `brew unbottled` would not tell me that it is `python@3.10` either. Does that make sense? I have to manually check the dependencies to see which one is unbottled, which seems like `brew unbottled` should be able to do.

2. for some reason, at least on linux, I cannot use `brew unbottled` for `arm64` systems:
```
$ brew unbottled --tag=arm64_monterey macaulay2
==> Populating dependency tree...
==> :arm64_monterey bottle status
macaulay2: doesn't support this macOS
```
**The problem:** The `macaulay2` formula is an example which definitely supports arm64_monterey and people have been able to bottle it after bottling a few dependencies, but I could not use `brew unbottled` to figure out which dependencies needed bottles from my Linux machine!

**Reply by mahrud** [reply]
I should add: `brew unbottled` and `brew test-bot` are both very useful tools and I appreciate your and other people's work on them. The above comments are not complaints but rather minor bug reports that if I was more familiar with the system I would try to fix myself, but since I can't, I hope explaining the issues above is useful to someone who can.

**Reply by SMillerDev** [reply]
I think both should be reported as bugs since that doesn't seem like expected behaviour.
